src/engine/Settings.cs

```diff
@@ -179,59 +285,55 @@ public void ApplySoundLevels()
         AudioServer.SetBusMute(music, VolumeMusicMuted);
     }
 
+    /// <summary>
+    ///   Applies current window settings to any applicable engine systems.
+    /// </summary>
     public void ApplyWindowSettings()
     {
         OS.WindowFullscreen = FullScreen;
         OS.VsyncEnabled = VSync;
     }
 
-    public void ApplyGraphicsSettings()
-    {
-        GUICommon.Instance.GetTree().Root.GetViewport().Msaa = MSAAResolution;
-        ColourblindScreenFilter.Instance.SetColourblindSetting(ColourblindSetting);
-    }
-
-    /// <summary>
-    ///   Reset all options to default values
-    /// </summary>
-    public void ResetToDefaults()
-    {
-        var defaults = new Settings();
-        var type = GetType();
-
-        foreach (var property in type.GetProperties())
-        {
-            if (!property.CanWrite)
-                continue;
-
-            property.SetValue(this, property.GetValue(defaults));
-        }
-    }
-
+    // Creates and returns a settings object loaded from the configuration settings file, or defaults if that fails.
     private static Settings LoadSettings()
-    {
-        var settings = LoadSettingsFileOrDefault();
-        settings.ApplyAll();
-        return settings;
-    }
-
-    private static Settings LoadSettingsFileOrDefault()
     {
         using (var file = new File())
         {
+            Settings settings;
```

Comment:
Isn't this variable definition basically redundant?

Reply:
Not sure what you mean by redundant. If you mean I could move the declaration of it to inside the error check and then just have the deserialize return the result right away, then yes you could do that instead.

Reply:
Move this variable definition to where it is needed.

Reply:
It is where it is needed, if I moved it inside the error check it would be out of scope, since it's used in two different places. I actually could have done that but after noticing my mistake I need it to call ApplyAll on.

Reply:
It's used in two places that I see:
```
                GD.Print("Settings configuration file is missing or unreadable, using default settings.");

                settings = new Settings();
                settings.Save();

                return settings;
```
and:
```
            settings = JsonConvert.DeserializeObject<Settings>(text);

            return settings;
```
I don't see why those need to share the same variable.
I'd do it like this:
```
                GD.Print("Settings configuration file is missing or unreadable, using default settings.");

                var settings = new Settings();
                settings.Save();

                return settings;
```
and:
```
            return JsonConvert.DeserializeObject<Settings>(text);
```

Reply:
Yes but that second part is now:

```
settings = JsonConvert.DeserializeObject<Settings>(text);
settings.ApplyAll();

return settings;
```
So that doesn't really work.

Reply:
Still I'd really like that the completely separate flows defined their own variables.

Reply:
This is the json load method. This should be safe to call without this calling settings.ApplyAll();
The settings loading method should call ApplyAll (that is called in the readonly static assignment).
I'm pretty sure that was in my original code, which is why I really, really prefer much smaller PRs as it's easier to review that they didn't break something.

Reply:
I can give them separate variables, I just thought the redundancy was disliked. I'm not really sure what you mean by your last comment though, it needs to call ApplyAll at some point to actually apply the loaded settings.

Reply:
It's old C style to define variables at the start of a function. C++ (and presumably also C#) good style is that variables are not defined before they are used, and go out of scope as soon as possible. With that style defining variable at the start of a function here (where it isn't needed as it isn't used outside the return places) is wrong. Instead it should be defined in each scope as soon as it's needed.

> it needs to call ApplyAll at some point to actually apply the loaded settings.

Yes, but don't call it in the json load method, because if someone wants to make a method ReloadFromFile method, it shouldn't also apply the freshly loaded settings. Instead call it in `LoadSettings`

Reply:
But that IS in LoadSettings? I didn't make it into two methods like it was before, am I supposed to? That is the only real difference from how it is now, before you called LoadSettings to initialize the singleton and then that called another method to get and return a settings object.
```
if (error != Error.Ok)
{
    GD.Print("Settings configuration file is missing or unreadable, using default settings.");

    var settings = new Settings();
    settings.ApplyAll();
    settings.Save();

    return settings;
}

var text = file.GetAsText();

file.Close();

var settings = JsonConvert.DeserializeObject<Settings>(text);
settings.ApplyAll();

return settings;
```
Amusingly enough this doesn't actually work either, I'm getting a compiler error for the var settings, that it is used in an enclosing scope. Even though the inner part is its own scope and it should be gone before it ever hit the bottom part.

Although if I split all this into two methods then it would remove that problem anyway.

Reply:
I'm probably misremembering, anyway making it two separate actions to "load Settings from json" and "prepare the global settings (loading them and applying them)" would be nice anyway.
Would it help for the duplicate variableness if you put the json convert outside the using block (passing just the text outside, the file doesn't need to exist for parsing the settings).